src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -623,6 +626,11 @@ private bool TakeStartLine(SocketInput input)
                     return false;
                 }
 
+                char[] requestUriChars;
+                var rawPathLength = beginPath.GetCharArray(endPath, out requestUriChars);
+                var decodedLength = UrlPathDecoder.DecodeInPlace(requestUriChars, rawPathLength);
```

Comment:
DecodeInPlace should take beginPath/endPath iterators, modify the bytes in-place to unescape % sequences, and return the new endPath position rather than copy to an intermediate char[]. This also means the DecodeInPlace method does not need to understand utf-8 sequences, because the subsequent conversion from byte[] to string via UTF8 encoder will take care of that concern.